### ISSUE TITLE
Add condition variables to the platform abstraction layer

### ISSUE THREAD

**Issue by yanesca** [Mbed-TLS/mbedtls#8455]
### Suggested enhancement

After some team discussion, we've decided to rely on a new threading abstraction which mimics C11 (i.e. mbedtls_fff where fff is the C11 function name, having the same parameters and return type, with default implementations for C11, pthreads and Windows). We'll use condition variables in addition to mutexes.

### Justification

Mbed TLS needs this because without it we can't implement the improvements to the threading support we were planning. (See the [threading design document](https://github.com/Mbed-TLS/mbedtls/blob/a365efc6f13e206d07732261b2156087091fad1c/docs/architecture/psa-thread-safety.md#destruction-of-a-key-in-use)).

**Comment by tom-cosgrove-arm** [comment]
I'd actually prefer that we move to reader-writer locks after mutexes

**Comment by yanesca** [comment]
I prefer condition variables as
- they are present in C11 threads - not because it would be necessary to translate 1-1 to any API, but in the hope that the C11 standards guys did a good job (they definitely put more thought into this than we can hope to) and this falls in the least common denominator for threading on a very wide range of platforms
- are more powerful - I mean I know how to implement reader-writer locks with condition variables, but don't know how to do it the way around

Also, if we decide to seriously consider reader-writer locks, we will need to plan/write up the design at least to the extent we have for condition variables.

**Comment by tom-cosgrove-arm** [comment]
RW locks are present in Pthreads and WIndows. As you say, they are a higher-level abstraction (can be built from mutexes and condition variables) so code size would be reduced and maintainability higher (i.e. don't use a low-level abstraction when there's a more appropriate one for the problem to be solved).

In this case, I think reader-writer locks do map well to what we're trying to do: once correct thread safety is implemented with mutexes, each mutex acquisition/release code could be replaced with either read or write acquire/release, and I would expect concurrency to increase and correctness remain.

**Comment by gilles-peskine-arm** [comment]
@paul-elliott-arm [Condition variables](https://en.wikipedia.org/wiki/Monitor_(synchronization)#Condition_variables) are more general than reader-writer locks (I don't think it's possible to implement condition variables with reader-writer locks without busy-wait). I think mutex plus condition variables are theoretically equivalent to semaphores, but semaphores are harder to use.

A typical use case is a producer-consumer queue: multiple producer threads can add objects to a queue, and multiple consumer threads can take object out of a queue. You can use a mutex to protect the queue, but how do consumers wake up producers? With a condition variable, the producer signals the condition, which wakes up consumers.

Condition variables are pretty common in concurrency APIs, and where they're missing there are usually semaphores. For example [Zephyr](https://docs.zephyrproject.org/latest/doxygen/html/group__condvar__apis.html) has condition variables, but if you want a reader-writer lock, you need to build your own.

**Comment by gilles-peskine-arm** [comment]
If you're curious about the choice of synchronization primitive, you might want to read [Jouni Leppäjärvi's thesis: *A pragmatic, historically oriented survey
on the universality of synchronization
primitives*](https://www.enseignement.polytechnique.fr/informatique/INF431/X09-2010-2011/AmphiTHC/SynchronizationPrimitives.pdf). Leppäjärvi presents (local) synchronization as being mutual exclusion plus wait-signal, which can be satisfactorily fulfilled by either semaphores or mutexes+condition variables.

**Comment by paul-elliott-arm** [comment]
I know what condition variables are, and I'm not questioning that we could bring them in, they are, as you say, pretty common. What I am saying here, is that in most cases we *need* something that looks very much almost in every way like a reader-writer lock (something, btw,  which is in every single multithreaded implementation I have ever worked with). 

I agree that we don't have RW locks on every platform, as I said above we should use them on platforms where we have them, and then do what we can for the ones that do not basically. You can construct a RW lock with care from a condition variable and a mutex.

So in short, I am saying we need both.

**Comment by yanesca** [comment]
As far as I remember the main argument for reader/writer locks was performance: taking a reader lock doesn't mean a full context switch if the OS implements the reader/writer lock. It takes a full context switch if we implement it on top of other primitives.

**Comment by gilles-peskine-arm** [comment]
We are very much pressed for time, but I'd still like to squeeze this into TF-PSA-Crypto 1.0. This means we have to minimize the amount of work, while still achieving the essential objective, which is is to have a concurrency abstraction that won't limit our options too much.

So what I propose to do is to only extend the current abstraction. No moving to C11. Just extend the existing pthread-like mutex interface with a pthread-like condition variable interface. As before, we would provide a full implementation when `MBEDTLS_THREADING_PTHREAD` is enabled, and for other platforms, it'll be up to the integrator.

Reader-writer locks are out as the only abstraction because they have limited power. They're what we wanted to have when we did the multithreaded key store in 3.6, but we there's a significant chance that we'll run into problems that they can't solve in the future. Given the limited time we have, they aren't going to go into 1.0. Later, we may add a rwlock module with a default implementation on top of mutex+cd, and the option to substitute an alternative implementation during platform integration (to gain performance or code size).

The work here will be:

* [DONE] Essentially no more design work. We were definitely not going to design a threading abstraction anyway, just pick one. I consider this done.
* [MUST] Extend `threading.h` with condition variable functions and their documentation.
* [MUST] Extend `threading.c` with the trivial wrappers for pthreads.
* [SHOULD] Write some smoke tests, just so that `make test` fails if you enable `MBEDTLS_THREADING_C` without providing cv functions.

**Comment by gilles-peskine-arm** [comment]
A couple of data points on C11 threads adoption.

In Glibc, it was added in [Glibc 2.18 released in August 2018](https://sourceware.org/legacy-ml/libc-alpha/2018-08/msg00003.html). In Ubuntu LTS, it's available since 20.04 (focal).

[ThreadSanitizer (TSan) doesn't support C11 threads as of 2025](https://github.com/google/sanitizers/issues/1603).

Nonetheless I think the ability to work on top of C11 threads is important, not directly because I expect a lot of people to use it, but because it's a feature set that more platform implementers have found acceptable than pthreads.

**Comment by gilles-peskine-arm** [comment]
We had a team discussion today, guided by the prototype in https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/415. Conclusions:

* Keep `mbedtls_threading_set_alt()`. It's not pretty but making the platform functions static causes a lot of problems under the hood.
* Add 5 parameters to `mbedtls_threading_set_alt()` for the condition variable functions.
* Add a level of indirection between `mbedtls_threading_mutex_t` and the type passed to the mutex functions.
* No impact on the mutex usage verification framework.
* No action now on adding thread create/join, we'll add that as a requirement later if you want to run multithreaded tests with `MBEDTLS_THREADING_ALT`.

**Comment by gilles-peskine-arm** [comment]
Looking more closely at [`mbedtls/threading.h`](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.6.4/include/mbedtls/threading.h), I think we need to do a few more minor changes that will require changes to the framework:

* Rename `MBEDTLS_ERR_THREADING_MUTEX_ERROR` (since it is now a mutex-or-condition-variable error) or replace it by a PSA error. (If `mbedtls_threading_condition_wait()` fails, there's no way to tell if the problem is due to the mutex or to the condition variable or to the combination.) This could map to `PSA_ERROR_BAD_STATE`, since an error (other than resource allocation) is likely a problem with the state of the mutex object or the state of the condition variable object. However, when a function fails due to an error in synchronization under the hood, it's probably useful to convey that the problem is related to concurrency rather than to the state of the domain of the function. So I think there should be a **new error name `MBEDTLS_ERR_THREADING_BAD_STATE`**.

* Inside the library, `MBEDTLS_ERR_THREADING_BAD_INPUT_DATA` is only used to report a null pointer. This is pretty much pointless (and arguably harmful because then static analyzers won't complain of passing a null pointer where it shouldn't be), so I don't plan to have null checks on the condition variable functions. So I plan to:
    1. Remove the null checks from the mutex functions.
    2. Change the framework to not require the library to define `MBEDTLS_ERR_THREADING_BAD_INPUT_DATA` (it's currently used to signal that thread create/join fails).
    3. Remove the now useless `MBEDTLS_ERR_THREADING_BAD_INPUT_DATA`.

* `threading.h` exposes the **function pointers for the mutex functions**, although they're undocumented. They should be in an internal header `threading_internal.h` instead. They are used in the mutex usage verification framework, which is in the framework repository. I don't plan to expose the condition variable function pointers, so for consistency we should remove the mutex variables too. This can be done after 1.0 since those identifiers aren't documented.

* `threading.h` exposes the **global mutexes**. I don't know why: they aren't used by test code or even by mbedtls, only by other crypto library functions.  They should move to `threading_internal.h` as well, or to their domain's header.

Only the new error code `MBEDTLS_ERR_THREADING_BAD_STATE` is necessary to write the condition variable code. Changing the existing errors would conflict with Note that this will have a minor conflict with https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/422 which makes `MBEDTLS_ERR_THREADING_BAD_INPUT_DATA` an alias for `PSA_ERROR_INVALID_ARGUMENT`. So my current plan is to do this in two steps:

1. Add the new error and the condition variables. PR to crypto only.
2. In 3.6 and crypto, create `threading_internal.h`, to prepare for the framework to start including it.
   https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/430 https://github.com/Mbed-TLS/mbedtls/pull/10375
4. Adapt the framework for the `threading.h` crud removal. This includes having it include `threading_internal.h`.
5. Do the crud removal in `threading.h` in crypto. Optionally move some stuff in 3.6 as well, but since it's an LTS, we probably shouldn't (even though application code really shouldn't be fiddling with the mutex functions or the global mutexes ).
